### PR TITLE
[#21] [Backend] As a user, I can search for the special characters in url of search result

### LINKED ISSUE #21: [Backend] As a user, I can search for the special characters in url of search result
## Why

User may need to know whether certain special characters are there in the url.

## Acceptance Criteria

- Make function to know regex search the url of search result for special characters. (URLs in stored search results with 2 or more / or 1 or more >).
- Add tests

### PR DESCRIPTION
**Pull request by Prabeshpd**
- Close #21 

## What happened 👀

- Add filter logic to search results for specific characters to occur at least n times in urls
- Add test

## Insight 📝

Added filter logic to get the search result for specific characters in urls. The logic is added in same query object for filtering with others.

-  In this scenario I have first pulled all the search results of user and applied filter on code logic.

## Proof Of Work 📹

Passing CI pipelines

### CODE REVIEW

File: src/queries/SearchResultQuery.ts [hunk 1]
```diff
@@ -17,21 +24,32 @@ interface SearchResultQueryInterface extends SearchResultQueryParams {
...
+    super({ ...queryParams, databaseModal: dbClient.searchResult });
 
-    this.keyword = searchResultQueryParams.keyword;
-    this.userId = searchResultQueryParams.userId;
+    this.userId = userId;
+    this.keyword = queryParams.filters?.keyword;
+    this.url = queryParams.filters?.url;
+    this.filterAdUrlWord = queryParams.filters?.filterAdUrlWord;
+    this.filterUrlCharacter = queryParams.filters?.filterUrlCharacter;
+    this.filterUrlCharacterAtLeast = queryParams.filters?.filterUrlCharacterAtLeast || 1;
```
Comment (tednguyendev) [inline src/queries/SearchResultQuery.ts]: These grow fast 🙈 We don't need to specify them like that, right?

```suggestion
    this.filters = queryParams.filters;
```

File: src/queries/SearchResultQuery.ts [hunk 2]
```diff
@@ -17,21 +24,32 @@ interface SearchResultQueryInterface extends SearchResultQueryParams {
...
     this.filterQuery = {};
   }
 
   async filterSearchResultList() {
     this.buildUserQueryParams();
     this.buildKeywordQueryParams();
     this.buildPaginationQueryParams();
+    this.buildUrlToFilterQueryParams();
+    await this.buildUrlWithWordFilterQueryParams();
+    await this.buildUrlContainsAtLeastSpecificCharactersQueryParams();
```
Comment (tednguyendev) [inline src/queries/SearchResultQuery.ts]: Same as [this](https://github.com/nimblehq/ic-nextjs-prabesh-dahal/pull/101/files#r1315698403)

File: src/queries/SearchResultQuery.ts [hunk 3]
```diff
@@ -48,6 +66,68 @@ class SearchResultQuery extends BaseQuery<SearchResult> implements SearchResultQ
...
+    // TODO: Remove after the process gets approved
+    // return dbClient.$executeRaw`
+    //   SELECT * FROM search_results sr
+    //    WHERE sr."userId" = ${this.userId}
+    //     AND array_to_string(sr.ad_word_top_urls, ', ') ~ ${this.filterAdUrlWord}
+    // `;
+  }
+
+  private async buildUrlContainsAtLeastSpecificCharactersQueryParams() {
+    if (!this.filterUrlCharacter) return;
```
Comment (tednguyendev) [inline src/queries/SearchResultQuery.ts:103]:
```suggestion
    if (!this.filterUrlCharacter) return;
    
```

Let's have a line break

File: src/queries/SearchResultQuery.ts [hunk 4]
```diff
@@ -48,6 +66,68 @@ class SearchResultQuery extends BaseQuery<SearchResult> implements SearchResultQ
...
+
+    // TODO: Remove after the process gets approved
+    // return dbClient.$executeRaw`
+    //   SELECT * FROM search_results sr
+    //    WHERE sr."userId" = ${this.userId}
+    //     AND array_to_string(sr.ad_word_top_urls, ', ') ~ ${this.filterAdUrlWord}
+    // `;
+  }
+
+  private async buildUrlContainsAtLeastSpecificCharactersQueryParams() {
```
Comment (tednguyendev) [inline src/queries/SearchResultQuery.ts:102]: 🙈

File: src/queries/SearchResultQuery.ts (hunk 2 above)
Comment (Prabeshpd) [reply src/queries/SearchResultQuery.ts]: Fixed in https://github.com/nimblehq/ic-nextjs-prabesh-dahal/pull/102/commits/10754243331fb6409ff0fac1798c474d7674afc1

File: src/queries/SearchResultQuery.ts [hunk 5]
```diff
@@ -30,6 +30,7 @@ class SearchResultQuery extends BaseQuery<SearchResult> implements SearchResultQ
     this.buildKeywordQueryParams();
     this.buildUrlQueryParams();
     await this.buildUrlWithWordQueryParams();
+    await this.buildUrlWithAtLeastCharactersQueryParams();
     this.buildPaginationQueryParams();
```
Comment (tednguyendev) [inline src/queries/SearchResultQuery.ts:34]: Let's have a line break to separate the filter builders and the pagination builders 💪 

```suggestion
    this.buildKeywordQueryParams();
    this.buildUrlQueryParams();
    await this.buildUrlWithWordQueryParams();
    await this.buildUrlWithAtLeastCharactersQueryParams();
    
    this.buildPaginationQueryParams();
```

File: test/queries/searchResultQuery.test.ts [hunk 6]
```diff
@@ -243,8 +242,8 @@ describe('SearchResultQuery', () => {
     });
   });
 
-  describe('given all valid filters', () => {
-    describe('given user has search results', () => {
+  describe('given special characters in url params', () => {
+    describe('given valid params for special characters in url filter', () => {
```
Comment (tednguyendev) [inline test/queries/searchResultQuery.test.ts:246]:
```suggestion
    describe('given there are special characters filters', () => {
```

What do you think? 👀

File: src/queries/SearchResultQuery.ts [hunk 7]
```diff
@@ -84,6 +85,29 @@ class SearchResultQuery extends BaseQuery<SearchResult> implements SearchResultQ
     //     AND array_to_string(sr.ad_word_top_urls, ', ') ~ ${this.filterAdUrlWord}
     // `;
   }
+
+  private async buildUrlWithAtLeastCharactersQueryParams() {
+    if (!this.filters?.filterUrlCharacter) return;
+
+    const characterToSearch = this.filters.filterUrlCharacter;
+    const leastCharacterCount = this.filters?.filterUrlCharacterAtLeast || 1;
```
Comment (tednguyendev) [inline src/queries/SearchResultQuery.ts:93]: [Ditto](https://github.com/nimblehq/ic-nextjs-prabesh-dahal/pull/101/files#r1316683006)

**Comment by Prabeshpd** [comment]
Closing this as the changes are in [PR](https://github.com/nimblehq/ic-nextjs-prabesh-dahal/pull/101)

**Pull request closed without merging**